### ISSUE TITLE
Usage of dbutil WithTxV inside standard adapter

### ISSUE THREAD

**Issue by ivan-penchev** [riverqueue/river#198]
Hi maintainers excellent job on this library!

I have two questions, so not an urgent issue, I tried to add a label to this issue but I guess it's not supported if you are not a maintainer.

In dbutil you have this function:
https://github.com/riverqueue/river/blob/ca5d0c44fccc244404b64f8a44f95293b12b78ee/internal/util/dbutil/db_util.go#L40

When I look at the standard adapter, I can see that sometimes it is used:
https://github.com/riverqueue/river/blob/ca5d0c44fccc244404b64f8a44f95293b12b78ee/internal/dbadapter/db_adapter.go#L522

And sometimes, the transaction is started manually, instead:
https://github.com/riverqueue/river/blob/ca5d0c44fccc244404b64f8a44f95293b12b78ee/internal/dbadapter/db_adapter.go#L460

Q1: I can't figure out the reason for this  inconsistent use, is it because of the default value for an array being nil if not explicitly given?
Meaning if we modify  WithTxV like so, it would avoid the code duplication?
```go
func WithTxV[T any](ctx context.Context, txBeginner TxBeginner, innerFunc func(ctx context.Context, tx pgx.Tx) (T, error)) (T, error) {
	var defaultRes T

	tx, err := txBeginner.Begin(ctx)

	// Use reflection to determine if T is a slice, map, or channel and initialize it appropriately.
	tType := reflect.TypeOf(defaultRes)
	switch tType.Kind() {
	case reflect.Slice:
		defaultRes = reflect.MakeSlice(tType, 0, 0).Interface().(T)
	case reflect.Map:
		defaultRes = reflect.MakeMap(tType).Interface().(T)
}
...//rest

```

Q2:
I probably am missing something somewhere, but is JobSetStateIfRunning not running in a transaction? And if yes, why?
It seems like it calls directly the queries, which in their turn just uses QueryRow via the executor?
I traced the function's use inside the executorer and completer, but it still doesn't grab a transaction.

Thank you very much for your work on this project. This is me trying to understand it a bit better, and not questioning you.

**Comment by brandur** [comment]
Hi Ivan,

> Q1: I can't figure out the reason for this inconsistent use, is it because of the default value for an array being nil if not explicitly given?

Don't read too much into this -- Blake wrote the bulk of the initial River code, and then I came in and started changing/adding stuff. `WithTxV` is a pattern that I'd used to good effect in other projects (it's more succinct, but another big benefit is that it makes it impossible to forget a `tx.Commit()`, which is a surprisingly common error), and I brought it in. I started using it in places, but we didn't go through and port it everywhere.

I don't think that universal use is even necessary. The normal `Begin`/`Commit` syntax isn't _that_ bad, and in some places it's necessary, like if you want to inject a transaction into another function, which will then commit it.

> Q2:
> I probably am missing something somewhere, but is JobSetStateIfRunning not running in a transaction? And if yes, why?
> It seems like it calls directly the queries, which in their turn just uses QueryRow via the executor?
> I traced the function's use inside the executorer and completer, but it still doesn't grab a transaction.

So if you think about the lifecycle of a job:

1. Locked by a `Producer`, and handed off to `jobExecutor`.
2. Worked by a user `Work` implementation.
3. Set as completed by a `Completer`.

`Work` implementations in (2) will often want to start their own transaction so they can work atomically, but River doesn't directly have access to that, so `Completer` can't complete jobs in it for the user.

It would be possible to open a transaction in the producer in (1), lock jobs, work them, have the completer in (3) set them as complete, then commit the whole transaction, but remember that the whole process is operating on a batch of jobs rather than a single one. The producer fetches as many jobs as it can, distributes them out for work, then goes for the next set. That's a lot of independent operations in one transaction where something could potentially go wrong and cause a rollback. I'd also argue that there isn't a huge amount of value to a transaction across (1) and (3) if you're not capturing the actual work in (2).

That's part of the rationale, but the other is that completing out of transaction is a performance optimization. The `Completer` currently spins up goroutines to run completions in parallel, but we intend to optimize this further. Think for example of a completer that delayed a little while (e.g. 10-50ms) before marking a job complete so that it could see if more completions came in. If they did, they'd all be done in a big batch query -- much more efficient, and saving many round trips to the database.

Users will want to complete a job atomically in some situations though, which is why River provides a facility for transactional job completion with `JobCompleteTx` that runs inside a work body:

https://riverqueue.com/docs/transactional-job-completion
